packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestApi.kt

```diff
@@ -0,0 +1,252 @@
+package com.amazonaws.amplify.amplify_api.rest_api
+
+import android.os.Handler
+import android.os.Looper
+import com.amazonaws.amplify.amplify_api.FlutterApiErrorMessage
+import com.amazonaws.amplify.amplify_api.FlutterApiError
+import com.amazonaws.amplify.amplify_api.OperationsManager
+import com.amplifyframework.api.ApiException
+import com.amplifyframework.api.rest.RestOperation
+import com.amplifyframework.api.rest.RestOptions
+import com.amplifyframework.api.rest.RestResponse
+import com.amplifyframework.core.Amplify
+import com.amplifyframework.core.Consumer
+import io.flutter.plugin.common.MethodChannel.Result
+import kotlin.reflect.KFunction3
+import kotlin.reflect.KFunction4
+
+typealias RestAPIFunction3 = KFunction3<
+        @ParameterName(name = "restOptions") RestOptions,
+        @ParameterName(name = "restConsumer") Consumer<RestResponse>,
+        @ParameterName(name = "exceptionConsumer") Consumer<ApiException>,
+        RestOperation?>
+
+typealias RestAPIFunction4 = KFunction4<
+        @ParameterName(name = "apiName") String,
+        @ParameterName(name = "restOptions") RestOptions,
+        @ParameterName(name = "restConsumer") Consumer<RestResponse>,
+        @ParameterName(name = "exceptionConsumer") Consumer<ApiException>,
+        RestOperation?>
+
+class FlutterRestApi {
+    companion object {
+        private val LOG = Amplify.Logging.forNamespace("amplify:flutter:api")
+
+        private fun restFunctionHelper(
+                methodName: String,
+                flutterResult: Result,
+                request: Map<String, *>,
+                function3: RestAPIFunction3,
+                function4: RestAPIFunction4
+        ) {
+            if (!FlutterRestInputs.isValid(flutterResult, request)) return
+
+            val inputs = FlutterRestInputs(request)
+            val cancelToken = inputs.getCancelToken()
+            val apiName: String? = inputs.getApiPath()
+            val options: RestOptions = FlutterRestInputs(request).getRestOptions()
+
+            try {
+
+                if (apiName == null) {
+                    var operation: RestOperation? = function3(options,
+                            Consumer { result -> prepareRestResponseResult(flutterResult, result, methodName, cancelToken) },
```

Comment:
we should remove the operation from the map here instead of in the `prepareRestResponseResult` as per single responsibility principle. Same in `prepareError`